nifi-nar-bundles/nifi-gcp-bundle/nifi-gcp-processors/src/main/java/org/apache/nifi/processors/gcp/drive/PutGoogleDrive.java

```diff
@@ -0,0 +1,495 @@
+            BASE_FOLDER_ID,
+            FOLDER_ID,
+            FOLDER_NAME,
+            FILE_NAME,
+            CONFLICT_RESOLUTION,
+            CREATE_FOLDER,
+            CHUNKED_UPLOAD_THRESHOLD,
+            CHUNKED_UPLOAD_SIZE,
+            ProxyConfiguration.createProxyConfigPropertyDescriptor(false, ProxyAwareTransportFactory.PROXY_SPECS)
+    ));
+
+    public static final Relationship REL_SUCCESS =
+            new Relationship.Builder()
+                    .name("success")
+                    .description("Files that have been successfully written to Google Drive are transferred to this relationship.")
+                    .build();
+
+    public static final Relationship REL_FAILURE =
+            new Relationship.Builder()
+                    .name("failure")
+                    .description("Files that could not be written to Google Drive for some reason are transferred to this relationship.")
+                    .build();
+
+    public static final Set<Relationship> RELATIONSHIPS = Collections.unmodifiableSet(new HashSet<>(asList(
+            REL_SUCCESS,
+            REL_FAILURE
+    )));
+
+    public static final String MULTIPART_UPLOAD_URL = "https://www.googleapis.com/upload/drive/v3/files?uploadType=multipart";
+
+    private volatile Drive driveService;
+    private Future<File> uploadedFileFuture;
+
+    @Override
+    public Set<Relationship> getRelationships() {
+        return RELATIONSHIPS;
+    }
+
+    @Override
+    public List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+        return PROPERTIES;
+    }
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        FlowFile flowFile = session.get();
+        if (flowFile == null) {
+            return;
+        }
+
+        String folderId = null;
+        final String filename = context.getProperty(FILE_NAME).evaluateAttributeExpressions(flowFile).getValue();
+        final String mimeType = flowFile.getAttribute(CoreAttributes.MIME_TYPE.key());
+        final String folderName = context.getProperty(FOLDER_NAME).evaluateAttributeExpressions(flowFile).getValue();
+
+        try {
+            folderId = createFoldersAndGetParentId(context, flowFile);
+            final long startNanos = System.nanoTime();
+            final ExecutorService uploadExecutor = Executors.newSingleThreadExecutor();
+            final long size = flowFile.getSize();
+
+            final long chunkUploadThreshold = context.getProperty(CHUNKED_UPLOAD_THRESHOLD)
+                    .asDataSize(DataUnit.B)
+                    .longValue();
+
+            final int uploadChunkSize = context.getProperty(CHUNKED_UPLOAD_SIZE)
+                    .asDataSize(DataUnit.B)
+                    .intValue();
+
+            final String conflictResolution = context.getProperty(CONFLICT_RESOLUTION).getValue();
+
+            Optional<File> alreadyExistingFile = checkFileExistence(filename, folderId);
```

Comment:
While it is a clean solution, it is also not that performant because, before every upload, there is another request to check whether the file exists. I am also afraid that the search behind the check might be slow. (Not sure about that.) In other cases (Azure, for example), we handle conflicts by catching the exceptions thrown by the upload part of the code. What do you think? Does it make sense here?

Reply:
You are right, that is the useful approach but it does not work in Google Drive case. In Google Drive you can create several folders, files with the same name. It does not throw error if you create az object with an already existing name, it happily creates another object with the same name (the id will be different, of course).